aptos-move/framework/move-stdlib/src/natives/bcs.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright © Aptos Foundation
+
+// Copyright (c) The Diem Core Contributors
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::natives::helpers::make_module_natives;
+use move_binary_format::errors::PartialVMResult;
+use move_core_types::{
+    gas_algebra::{InternalGas, InternalGasPerByte, NumBytes},
+    vm_status::sub_status::NFE_BCS_SERIALIZATION_FAILURE,
+};
+use move_vm_runtime::native_functions::{NativeContext, NativeFunction};
+use move_vm_types::{
+    loaded_data::runtime_types::Type,
+    natives::function::NativeResult,
+    pop_arg,
+    values::{values_impl::Reference, Value},
+};
+use smallvec::smallvec;
+use std::{collections::VecDeque, sync::Arc};
+/***************************************************************************************************
+ * native fun to_bytes
+ *
+ *   gas cost: size_of(val_type) * input_unit_cost +        | get type layout
+ *             size_of(val) * input_unit_cost +             | serialize value
+ *             max(size_of(output), 1) * output_unit_cost
+ *
+ *             If any of the first two steps fails, a partial cost + an additional failure_cost
+ *             will be charged.
+ *
+ **************************************************************************************************/
+#[derive(Debug, Clone)]
+pub struct ToBytesGasParameters {
+    pub per_byte_serialized: InternalGasPerByte,
+    pub legacy_min_output_size: NumBytes,
+    pub failure: InternalGas,
+}
+
+/// Rust implementation of Move's `native public fun to_bytes<T>(&T): vector<u8>`
+#[inline]
+fn native_to_bytes(
+    gas_params: &ToBytesGasParameters,
+    context: &mut NativeContext,
```

Comment:
I wonder if we should also start using SafeNativeContext in all of these functions. cc @vgao1996

Reply:
That is one reason why this is happening... perhaps in a different PR.

Reply:
I think it's good to separate the copying from the follow-ups of changing the code